In [1]:
import pandas as pd

In [2]:
crime_data = pd.read_csv('/Users/josepalacios/Documents/Machine Learning/boston_crime')

In [4]:
# Get the number of rows and columns
print(crime_data.shape)

(319073, 17)


In [5]:
# Get Column names
print(crime_data.columns)

Index(['INCIDENT_NUMBER', 'OFFENSE_CODE', 'OFFENSE_CODE_GROUP',
       'OFFENSE_DESCRIPTION', 'DISTRICT', 'REPORTING_AREA', 'SHOOTING',
       'OCCURRED_ON_DATE', 'YEAR', 'MONTH', 'DAY_OF_WEEK', 'HOUR', 'UCR_PART',
       'STREET', 'Lat', 'Long', 'Location'],
      dtype='object')


In [6]:
# Get the dtype of each Column
print(crime_data.dtypes)

INCIDENT_NUMBER         object
OFFENSE_CODE             int64
OFFENSE_CODE_GROUP      object
OFFENSE_DESCRIPTION     object
DISTRICT                object
REPORTING_AREA          object
SHOOTING                object
OCCURRED_ON_DATE        object
YEAR                     int64
MONTH                    int64
DAY_OF_WEEK             object
HOUR                     int64
UCR_PART                object
STREET                  object
Lat                    float64
Long                   float64
Location                object
dtype: object




| Pandas Type  | Python type | Description |
| --- | --- | --- |
| object | string | most common type |
| int64 | int | Whole Numbers |
| float64 | float | Numbers and 

In [7]:
# Get the first Row. 
print(crime_data.loc[0])

INCIDENT_NUMBER                         I182070945
OFFENSE_CODE                                   619
OFFENSE_CODE_GROUP                         Larceny
OFFENSE_DESCRIPTION             LARCENY ALL OTHERS
DISTRICT                                       D14
REPORTING_AREA                                 808
SHOOTING                                       NaN
OCCURRED_ON_DATE               2018-09-02 13:00:00
YEAR                                          2018
MONTH                                            9
DAY_OF_WEEK                                 Sunday
HOUR                                            13
UCR_PART                                  Part One
STREET                                  LINCOLN ST
Lat                                        42.3578
Long                                      -71.1394
Location               (42.35779134, -71.13937053)
Name: 0, dtype: object


In [9]:
print(crime_data.head(1))

  INCIDENT_NUMBER  OFFENSE_CODE OFFENSE_CODE_GROUP OFFENSE_DESCRIPTION  \
0      I182070945           619            Larceny  LARCENY ALL OTHERS   

  DISTRICT REPORTING_AREA SHOOTING     OCCURRED_ON_DATE  YEAR  MONTH  \
0      D14            808      NaN  2018-09-02 13:00:00  2018      9   

  DAY_OF_WEEK  HOUR  UCR_PART      STREET        Lat       Long  \
0      Sunday    13  Part One  LINCOLN ST  42.357791 -71.139371   

                      Location  
0  (42.35779134, -71.13937053)  


In [11]:
# Count Missing Values
crime_data.isnull().sum()

INCIDENT_NUMBER             0
OFFENSE_CODE                0
OFFENSE_CODE_GROUP          0
OFFENSE_DESCRIPTION         0
DISTRICT                 1765
REPORTING_AREA              0
SHOOTING               318054
OCCURRED_ON_DATE            0
YEAR                        0
MONTH                       0
DAY_OF_WEEK                 0
HOUR                        0
UCR_PART                   90
STREET                  10871
Lat                     19999
Long                    19999
Location                    0
dtype: int64

In [17]:
# Missing Values in one Column.
crime_data.SHOOTING.value_counts(dropna = False)

NaN    318054
Y        1019
Name: SHOOTING, dtype: int64

In [20]:
# Verify if the values filled are Y
crime_data.SHOOTING.value_counts()['Y']


1019

Shooting is boolean, Y or N . I will assume all the NaN = 318054 are N values Missings and replace empty with N. 

In [26]:
# Fill the missing values with 'N'
# Replace 'N' and 'Y' with False and True.
crime_data.SHOOTING.fillna('N', inplace= True)
crime_data.SHOOTING.replace({'Y': True, 'N': False}, inplace = True)

In [27]:
# Verify there are not missing values anymore. 
crime_data.SHOOTING.value_counts(dropna = False)


False    318054
True       1019
Name: SHOOTING, dtype: int64

In [30]:
# Check for duplicates 
# Lets assume Incident Number is the primary key 
crime_data.INCIDENT_NUMBER.value_counts()

I162030584    13
I152080623    11
I182065208    10
I172013170    10
I172096394    10
              ..
I152070889     1
I172008161     1
I172019021     1
I162104830     1
I162013762     1
Name: INCIDENT_NUMBER, Length: 282517, dtype: int64

In [35]:
# Lets analize duplicate values. 

crime_data[crime_data.INCIDENT_NUMBER == 'I162030584'].head(3).T

237660                       237661  \
INCIDENT_NUMBER                       I162030584                   I162030584   
OFFENSE_CODE                                 423                          802   
OFFENSE_CODE_GROUP            Aggravated Assault               Simple Assault   
OFFENSE_DESCRIPTION         ASSAULT - AGGRAVATED     ASSAULT SIMPLE - BATTERY   
DISTRICT                                     C11                          C11   
REPORTING_AREA                               243                          243   
SHOOTING                                   False                        False   
OCCURRED_ON_DATE             2016-04-20 11:07:00          2016-04-20 11:07:00   
YEAR                                        2016                         2016   
MONTH                                          4                            4   
DAY_OF_WEEK                            Wednesday                    Wednesday   
HOUR                                          11                           11   
UCR_PART                                Part One                     Part Two   
STREET                               TRESCOTT ST                  TRESCOTT ST   
Lat                                      42.3158                      42.3158   
Long                                    -71.0605                     -71.0605   
Location             (42.31580044, -71.06052053)  (42.31580044, -71.06052053)   

                                                            237662  
INCIDENT_NUMBER                                         I162030584  
OFFENSE_CODE                                                  1501  
OFFENSE_CODE_GROUP                              Firearm Violations  
OFFENSE_DESCRIPTION  WEAPON - FIREARM - CARRYING / POSSESSING, ETC  
DISTRICT                                                       C11  
REPORTING_AREA                                                 243  
SHOOTING                                                     False  
OCCURRED_ON_DATE                               2016-04-20 11:07:00  
YEAR                                                          2016  
MONTH                                                            4  
DAY_OF_WEEK                                              Wednesday  
HOUR                                                            11  
UCR_PART                                                  Part Two  
STREET                                                 TRESCOTT ST  
Lat                                                        42.3158  
Long                                                      -71.0605  
Location                               (42.31580044, -71.06052053)

The offense are different but the locations are the same. occured_on_date is the same, **there is a discrepancy**